source/base/parameter_acceptor.cc

```diff
@@ -24,25 +24,47 @@
 DEAL_II_NAMESPACE_OPEN
 
 
-// Static empty class list
-std::vector<SmartPointer<ParameterAcceptor>> ParameterAcceptor::class_list;
+namespace internal
+{
+  struct ParameterAcceptorCompare
+  {
+    bool
+    operator()(const ParameterAcceptor *p1, const ParameterAcceptor *p2) const
+    {
+      return p1->get_acceptor_id() < p2->get_acceptor_id();
+    }
+  };
+} // namespace internal
+
+
+// Static mutex for the class list
+std::mutex ParameterAcceptor::class_list_mutex;
+
+// Static empty class set
+std::set<ParameterAcceptor *, internal::ParameterAcceptorCompare>
+  ParameterAcceptor::class_list;
+
 // Static parameter handler
 ParameterHandler ParameterAcceptor::prm;
 
 ParameterAcceptor::ParameterAcceptor(const std::string &name)
-  : acceptor_id(class_list.size())
+  : acceptor_id(
+      class_list.empty() ? 0 : (*(class_list.rbegin()))->get_acceptor_id() + 1)
```

Comment:
This isn't strictly wrong (though it is not thread-safe -- but that's ok). But really, the idiom you're using here is that you want to get a new id every time you get to this point. You do this by asking the last object and increment the counter, but it's just as easy to write this with a function such as this (which is now thread-safe):
```
int next_free_id () {
  static std::mutex m;
  std::lock_guard l(m);

  static int current_id = 0;
  return current_id++;
}
```

Reply:
At least if you don't plan on ever creating more than 4B such objects.